marlowe-dashboard-client/src/Contract/State.purs

```diff
@@ -130,16 +140,65 @@ mkInitialState walletDetails currentSlot followerAppId { chParams, chHistory } =
+-- we don't have a design for displaying them anywhere, and because we are currently only
+-- using one in a special case (in the Escrow with Collateral contract), where it doesn't
+-- make much sense to show it to the user anyway. If we ever want to use PK parties for
+-- other purposes, we will need to rethink this.
+-- Note 2: In general there is no way to map parties to wallet nicknames. It is possible
+-- for the user who created the contract and distributed the role tokens, but this
+-- information will be lost when the browser is closed. And other participants don't have
+-- this information in the first place. Also, in the future it could be possible to give
+-- role tokens to other wallets, and we wouldn't know about that either. Still, we're
+-- keeping the `Maybe WalletNickname` in here for now, in case a way of making it generally
+-- available (e.g. through the metadata server) ever becomes apparent.
+getParticipants :: Contract -> Map Party (Maybe WalletNickname)
+getParticipants contract = Map.fromFoldable $ map (\x -> x /\ Nothing) (getRoleParties contract)
+
+getRoleParties :: Contract -> Array Party
+getRoleParties contract = filter isRoleParty $ Set.toUnfoldable $ getParties contract
+  where
+  isRoleParty party = case party of
+    Role _ -> true
+    _ -> false
+
+updateState :: WalletDetails -> MarloweParams -> Slot -> Array TransactionInput -> State -> State
+updateState walletDetails marloweParams currentSlot transactionInputs state =
   let
     previousTransactionInputs = toArrayOf (_executionState <<< _previousTransactions) state
 
     newTransactionInputs = difference transactionInputs previousTransactionInputs
 
+    setMarloweParams = set _mMarloweParams (Just marloweParams)
+
+    setUserParties = set _userParties (getUserParties walletDetails marloweParams)
+
+    clearPendingTransaction = set _pendingTransaction Nothing
+
     updateExecutionState = over _executionState (applyTransactionInputs newTransactionInputs)
+
+    -- If the `MarloweParams` are `Nothing`, that means this is the first update we've received for
+    -- a placeholder contract, so we need to set the `MarloweParams` now and also work out the
+    -- `userParties` (because these depend on the `MarloweParams`). Separately, if there are new
+    -- transaction inputs to apply, we need to clear the `pendingTransaction`. If we wanted to be
+    -- really careful, we could check that the new transaction input matches the pending one, but
+    -- I can't see how it wouldn't (and I don't think it matters anyway).
+    baseState = case isNothing $ state ^. _mMarloweParams, newTransactionInputs /= mempty of
+      true, true ->
+        state
+          # setMarloweParams
+          # setUserParties
+          # clearPendingTransaction
+      true, false ->
+        state
+          # setMarloweParams
+          # setUserParties
+      false, true ->
+        state
+          # clearPendingTransaction
+      false, false -> state
```

Comment:
Now that I see the comment I understand better what is going on, and it seems that the marlowe params modifications and the new transaction modifications are independent of each other. If there are pending transactions we need to `clearPendingTransaction` and if this is the first time we see marlowe params  we set them and the user parties...

I think the following would be a little clearer, but is totally up to you (with the current comment is clear enough).

```purescript
let
  mSetParamsAndParties = if isNothing (state ^. _mMarloweParams) 
    then 
       setMarloweParams >>> setUserParties 
    else 
       identity

  mClearPendingTransaction = if newTransactionInputs /= mempty 
     then 
        clearPendingTransaction 
     else 
        identity
in 
  state
     # mSetParamsAndParties
     # mClearPendingTransaction
     # updateExecutionState
     # regenerateStepCards currentSlot
     # selectLastStep
```

Reply:
I like it :+1: